Make test_torchbind.py training IR compatible

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139511
* #139233
* #139209
* __->__ #138658

In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object.
3. We don't run any DCE on training IR so we need to add some more torch ops to verifier. 

Differential Revision: [D64853530](https://our.internmc.facebook.com/intern/diff/D64853530)

tugsbayasgalan committed:
Make test_torchbind.py training IR compatible

[ghstack-poisoned]

ydwu4 commented on test/export/test_torchbind.py:
```diff
@@ -745,7 +754,33 @@ def forward(self, L_safe_obj_ : torch.ScriptObject):
+    return (getitem, sin)""",  # noqa: B950
+            )
+
+    @unittest.expectedFailure  # T205481814
```
Not sure what's going on here. Other parts looks good.

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object. 




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object. 




[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object.

Differential Revision: [D64853530](https://our.internmc.facebook.com/intern/diff/D64853530)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object.

Differential Revision: [D64853530](https://our.internmc.facebook.com/intern/diff/D64853530)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object.

Differential Revision: [D64853530](https://our.internmc.facebook.com/intern/diff/D64853530)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan commented on torch/_export/serde/serialize.py:
```diff
@@ -183,6 +183,7 @@ def _reverse_map(d: Dict[Any, Enum]):
     torch.sym_max,
     torch.sym_min,
     torch.sym_sqrt,
+    torch.fx.experimental.symbolic_shapes.cast_symbool_to_symint_guardless,
```
cc: @angelayi @zhxchen17 What do you think about this? Because training IR doesn't run any DCE, we end up with some ops like this that is related to deferred runtime assertion. Is this ok way to serialize? Or should it be a more "public" API?

zhxchen17 replied:
I don't think it should be allowed in the current form. What's the semantics of this? Do you expect this op to cause some confusion to the downstream?

pianpwk replied:
I think in some cases a boolean condition (e.g. `x = y in z`) ends up being represented with an unbacked symint with range [0, 1], and this is used for conversion from symbool -> symint.

I think it being in this list is valid, as it's a op producing a symint. But agree with @zhxchen17 that I have no idea how it'll be handled downstream.

tugsbayasgalan replied:
Hmm how do we expect to handle other ops producing symint in downstream today?

pianpwk replied:
not really sure. but since this list is just an attempt to enumerate ops that produce symints, I think it's fine? It would be nice to have a more generic check

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object.
3. We don't run any DCE on training IR so we need to add some more torch ops to the serializer. 

Differential Revision: [D64853530](https://our.internmc.facebook.com/intern/diff/D64853530)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object.
3. We don't run any DCE on training IR so we need to add some more torch ops to the serializer. 

Differential Revision: [D64853530](https://our.internmc.facebook.com/intern/diff/D64853530)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

pianpwk commented on torch/export/_trace.py:
```diff
@@ -1832,6 +1841,8 @@ def _export_for_training(
 
     from torch._export.verifier import TrainingIRVerifier
 
+    print("GRAPH", gm.graph)
```
nit

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object.
3. We don't run any DCE on training IR so we need to add some more torch ops to the serializer. 

Differential Revision: [D64853530](https://our.internmc.facebook.com/intern/diff/D64853530)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object.
3. We don't run any DCE on training IR so we need to add some more torch ops to the serializer. 

Differential Revision: [D64853530](https://our.internmc.facebook.com/intern/diff/D64853530)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object.
3. We don't run any DCE on training IR so we need to add some more torch ops to the serializer. 

Differential Revision: [D64853530](https://our.internmc.facebook.com/intern/diff/D64853530)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan committed:
Update on "Make test_torchbind.py training IR compatible"


In this diff, i make test_torchbind.py tests to handle training IR. Today in the training IR, we don't see the effect token and HOP because this happens at the FunctionalTensorMode. Maybe in the future, we should move this logic up to the training IR so that writing passes etc on training Ir is safer. But for the migration purposes, i think it is ok for now.  I also fixed two bugs: 
1. ep.module() doesn't register all aliased constants in the module.
2. When we retrace, we need to fakify the original Torchbind object.
3. We don't run any DCE on training IR so we need to add some more torch ops to the serializer. 

Differential Revision: [D64853530](https://our.internmc.facebook.com/intern/diff/D64853530)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64853530).

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -1071,7 +1071,17 @@ def forward(self, x, y):
+        # as input, runtime assertion should fail. This is because we would create
+        # guard on y.shape[0] > x.shape[0] but somehow in old export, we dce this
+        # assertion.
+        if is_training_ir_test(self._testMethodName) and is_non_strict_test(
```
cc: @pianpwk
